hazelcast/src/test/java/com/hazelcast/client/protocol/EnumCompatibilityTest.java

```diff
@@ -138,52 +138,53 @@ public void testProtocolType() {
     public void testFieldKind() {
         // Used in FieldDescriptorCodec
         Map<FieldKind, Integer> mappings = new HashMap<>();
-        mappings.put(FieldKind.BOOLEAN, 0);
-        mappings.put(FieldKind.ARRAY_OF_BOOLEAN, 1);
-        mappings.put(FieldKind.INT8, 2);
-        mappings.put(FieldKind.ARRAY_OF_INT8, 3);
-        mappings.put(FieldKind.CHAR, 4);
-        mappings.put(FieldKind.ARRAY_OF_CHAR, 5);
-        mappings.put(FieldKind.INT16, 6);
-        mappings.put(FieldKind.ARRAY_OF_INT16, 7);
-        mappings.put(FieldKind.INT32, 8);
-        mappings.put(FieldKind.ARRAY_OF_INT32, 9);
-        mappings.put(FieldKind.INT64, 10);
-        mappings.put(FieldKind.ARRAY_OF_INT64, 11);
-        mappings.put(FieldKind.FLOAT32, 12);
-        mappings.put(FieldKind.ARRAY_OF_FLOAT32, 13);
-        mappings.put(FieldKind.FLOAT64, 14);
-        mappings.put(FieldKind.ARRAY_OF_FLOAT64, 15);
-        mappings.put(FieldKind.STRING, 16);
-        mappings.put(FieldKind.ARRAY_OF_STRING, 17);
-        mappings.put(FieldKind.DECIMAL, 18);
-        mappings.put(FieldKind.ARRAY_OF_DECIMAL, 19);
-        mappings.put(FieldKind.TIME, 20);
-        mappings.put(FieldKind.ARRAY_OF_TIME, 21);
-        mappings.put(FieldKind.DATE, 22);
-        mappings.put(FieldKind.ARRAY_OF_DATE, 23);
-        mappings.put(FieldKind.TIMESTAMP, 24);
-        mappings.put(FieldKind.ARRAY_OF_TIMESTAMP, 25);
-        mappings.put(FieldKind.TIMESTAMP_WITH_TIMEZONE, 26);
-        mappings.put(FieldKind.ARRAY_OF_TIMESTAMP_WITH_TIMEZONE, 27);
-        mappings.put(FieldKind.COMPACT, 28);
-        mappings.put(FieldKind.ARRAY_OF_COMPACT, 29);
-        mappings.put(FieldKind.PORTABLE, 30);
-        mappings.put(FieldKind.ARRAY_OF_PORTABLE, 31);
-        mappings.put(FieldKind.NULLABLE_BOOLEAN, 32);
-        mappings.put(FieldKind.ARRAY_OF_NULLABLE_BOOLEAN, 33);
-        mappings.put(FieldKind.NULLABLE_INT8, 34);
-        mappings.put(FieldKind.ARRAY_OF_NULLABLE_INT8, 35);
-        mappings.put(FieldKind.NULLABLE_INT16, 36);
-        mappings.put(FieldKind.ARRAY_OF_NULLABLE_INT16, 37);
-        mappings.put(FieldKind.NULLABLE_INT32, 38);
-        mappings.put(FieldKind.ARRAY_OF_NULLABLE_INT32, 39);
-        mappings.put(FieldKind.NULLABLE_INT64, 40);
-        mappings.put(FieldKind.ARRAY_OF_NULLABLE_INT64, 41);
-        mappings.put(FieldKind.NULLABLE_FLOAT32, 42);
-        mappings.put(FieldKind.ARRAY_OF_NULLABLE_FLOAT32, 43);
-        mappings.put(FieldKind.NULLABLE_FLOAT64, 44);
-        mappings.put(FieldKind.ARRAY_OF_NULLABLE_FLOAT64, 45);
+        mappings.put(FieldKind.NOT_AVAILABLE, 0);
```

Comment:
Why do we put NOT_AVAILABLE to 0? It could be 46 as well. It won't be compatible with old versioned clients (of course it does not have to be because it is beta)

Do we really want to break compat due to this?

Reply:
It can certainly be 46, but I feel like this is a unique field kind, and the id of `0` seems like a good fit for it.

Honestly, I don't care about compatibility with 5.1 clients. They might work with 5.2 clusters by chance before this, but we didn't guarantee anything about it. If someone wants to use 5.2 members with Compact, they must use 5.2 clients.

@ihsandemir @yuce What do you guys think about the compatibility aspect of it?

Reply:
I am not too strong about this but having that compatibility may be a nice-to-have.

Reply:
I also do not worry about compatibility of Beta. Let's decide on what looks the best. Putting at the end is a problem if you add new types, hence I would also prefer putting at the front index of 0 as well.

Reply:
I agree with @srknzl here. I don't like the idea of setting `NOT_AVAILABLE == 0` and shifting all other values. This is a client-side only constant, no need to make it the same for all clients. And this change will introduce incompatibility for no benefit.

Reply:
I can make it `-1`, and not shift others if you want. But I don't believe we should make it `46`

Reply:
I think `-1` is better than `0` in this case.

Reply:
Can we also start the values for the new data types from `47` and keep `46` as a reserved value?

Reply:
Why would we do that? It will alter the logic of `FieldOperations#isArrayKind` (we would either need to reserve 47 as well, or alter the code so that it behaves when the id < 46 and id > 47, which is not nice)